sqlccl: preserve/discard FKs in RESTORE

dt opened the pull request:
if, and only if, both tables are restored do we preserve the FK.

If only a referenced table is restored, we remove the backref -- since we are a *new* table during restore, we know no existing tables reference us unless they are also in the restore.

If a referencing table is being restored without the table it references, we potentially could update the FK to point to an existing table with the same db and name if one exists, but mark it as unvalidated, though it is unclear that RESTORE should have side-effects on tables not restored, at least by default. Either way, this is left for a later followup.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14159)
<!-- Reviewable:end -->

dt committed:
sqlccl: preserve/discard FKs in RESTORE

if, and only if, both tables are restored do we preserve the FK.

If only a referenced table is restored, we remove the backref -- since we are a *new* table during restore, we know no existing tables reference us unless they are also in the restore.

If a referencing table is being restored without the table it references, we potentially could update the FK to point to an existing table with the same db and name if one exists, but mark it as unvalidated, though it is unclear that RESTORE should have side-effects on tables not restored, at least by default. Either way, this is left for a later followup.

dt committed:
sqlccl: switch some tests to singleNode

where we’re just looking at descriptor handling, multi-vs-single node
is uninteresting and the extra nodes just make the test take longer to start.

benesch commented:
Sorry for the late review. The approach here <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>, but see note below about this behavior being surprising.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/ccl/sqlccl/restore.go, line 239 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14159#-KfMm3SX84aNYGQ4gcU4:-KfMm3SX84aNYGQ4gcU5:brsj46l) ([raw file](https://github.com/cockroachdb/cockroach/blob/0edc68a24fbc06ad34be65c1307bbcc9abf02376/pkg/ccl/sqlccl/restore.go#L239)):*
> ```Go
> 					// a db and name matching the one the FK pointed to at backup, should
> 					// we update the FK to point to it?
> 					index.ForeignKey = sqlbase.ForeignKeyReference{}
> ```

I think I'd find this behavior pretty surprising, given that real schemas often have small join tables that are easy to forget about (e.g. `receipt_type` or `country_codes`). How would you feel about making this an error except in the presence of something like a `SKIP_MISSING_FOREIGN_KEYS` KV option?

If this was already specced out somewhere, feel free to ignore me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14159#-:-KfMnVrwNpqvu4mcmY5m:baug69s)*
<!-- Sent from Reviewable.io -->

dt committed:
sqlccl: default to err if restore missing referenced table

the skip_missing_foreign_keys option allows just restoring some table even without the tables it references.

dt commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/ccl/sqlccl/restore.go, line 239 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14159#-KfMm3SX84aNYGQ4gcU4:-KfN1blSzq-s4F-4SeHg:b-qm3tjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0edc68a24fbc06ad34be65c1307bbcc9abf02376/pkg/ccl/sqlccl/restore.go#L239)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I think I'd find this behavior pretty surprising, given that real schemas often have small join tables that are easy to forget about (e.g. `receipt_type` or `country_codes`). How would you feel about making this an error except in the presence of something like a `SKIP_MISSING_FOREIGN_KEYS` KV option?

If this was already specced out somewhere, feel free to ignore me.
</blockquote></details>

Yeah, I think there are ~3 behaviors for restoring tables with FKs to tables not in the restore that could make sense in various situations:
1) elide the FKs but allow restore to succeed
2) fail with an error saying referenced tables isn't in restore
3) attempt to find referenced tables in the restoring cluster.

Behavior 3) is a little iffy: the table, or columns referenced might not exist or might have been renamed or may not have an index on them anymore, plus is feels a little weird that restoring a table alters other tables.

Anyway, I originally thought we should add more behavior in a followup and simply getting the invalid IDs out of restored tables was a good start until we gave more thought to these various semantics, but since you, dan, and I all thought it was strange to silently kill them, I've gone ahead and added 2) as the default with an option.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14159)*
<!-- Sent from Reviewable.io -->

benesch commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>!

---

Reviewed 2 of 2 files at r3.
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14159#-:-KfNti1_96dm7SBJKPUp:bg7igkw)*
<!-- Sent from Reviewable.io -->
